File: WoofIOS/SwiftPackages/Data/DataSource/Sources/DataSourceFakes/Account/FakeAccountDataSource.swift
import Combine
import DI
import DTO
import DataSource
import Foundation
import Utils

@MockApi
public final class FakeAccountDataSource: AccountDataSourceImplementing {
    public var account = DogProfileDTO(
        dog: DogDTO(id: 0, name: "Milo", photoUrl: "woof://dogs/border_collie_01.jpg"),
        breed: "Border Collie",
        ageInYears: 3,
        size: DogSizeDTO.medium,
        neighborhood: "Schöneberg, Berlin",
        personality: ["Frisbee pro", "Early riser"],
        favoriteActivity: "Agility courses",
        bio: "Herds tennis balls for a living."
    )
    public var getAccountError: DataSourceError?

    private let scheduler: SchedulerProviding

    public func getAccount() -> AnyPublisher<DogProfileDTO, DataSourceError> {
        Deferred { [weak self] () -> AnyPublisher<DogProfileDTO, DataSourceError> in
            guard let self else { return Empty().eraseToAnyPublisher() }
            if let getAccountError {
                return Fail(error: getAccountError).eraseToAnyPublisher()
            }
            return Just(account).setFailureType(to: DataSourceError.self).eraseToAnyPublisher()
        }
        .receive(on: scheduler.mainScheduler)
        .eraseToAnyPublisher()
    }
}

File: WoofAndroid/data/datasource/src/testFixtures/kotlin/com/perrystreet/woof/datasource/account/FakeAccountDataSource.kt
package com.perrystreet.woof.datasource.account

import com.perrystreet.woof.dto.dog.DogDTO
import com.perrystreet.woof.dto.dog.DogProfileDTO
import com.perrystreet.woof.dto.dog.DogSizeDTO
import io.reactivex.rxjava3.core.Single

@org.koin.core.annotation.Single
class FakeAccountDataSource : IAccountDataSource {
    var account: DogProfileDTO = DogProfileDTO(
        dog = DogDTO(id = 0, name = "Milo", photoUrl = "file:///android_asset/dogs/border_collie_01.jpg"),
        breed = "Border Collie",
        ageInYears = 3,
        size = DogSizeDTO.Medium,
        neighborhood = "Schöneberg, Berlin",
        personality = listOf("Frisbee pro", "Early riser"),
        favoriteActivity = "Agility courses",
        bio = "Herds tennis balls for a living.",
    )
    var getAccountError: Throwable? = null

    override fun getAccount(): Single<DogProfileDTO> =
        getAccountError?.let { Single.error(it) } ?: Single.just(account)
}
